File: typescript/src/lib/bulk_command_execution/v0_bulk_command_api/v0_command_types.ts
import { ok, err, Result } from "neverthrow";

// We provide a visitor interface here so that:
//  1) all enum cases can be exhaustively handled
//  2) any changes in the enum will result in a compile break
export interface V0CommandTypeVisitor {
	visitLoadLambda: () => Result<null, Error>;
	visitExecuteLambda: () => Result<null, Error>;
	visitRegisterService: () => Result<null, Error>;
	visitGenerateFiles: () => Result<null, Error>;
	visitLoadStaticFiles: () => Result<null, Error>;
	visitStartService: () => Result<null, Error>;
	visitRemoveService: () => Result<null, Error>;
	visitRepartition: () => Result<null, Error>;
	visitExecCommand: () => Result<null, Error>;
	visitWaitForHttpGetEndpointAvailability: () => Result<null, Error>;
	visitWaitForHttpPostEndpointAvailability: () => Result<null, Error>;
	visitExecuteBulkCommands: () => Result<null, Error>;
	visitGetServices: () => Result<null, Error>;
	visitGetLambdas: () => Result<null, Error>;
}

// vvvvvvvvvvvvvvvvvvvv Update the visitor whenever you add an enum value!!! vvvvvvvvvvvvvvvvvvvvvvvvvvv
export enum V0CommandType {
	LoadLambda = "LOAD_LAMBDA",
	ExecuteLambda = "EXECUTE_LAMBDA",
	RegisterService = "REGISTER_SERVICE",
	GenerateFiles = "GENERATE_FILES",
	LoadStaticFiles = "LOAD_STATIC_FILES",
	StartService = "START_SERVICE",
	RemoveService = "REMOVE_SERVICE",
	Repartition = "REPARTITION",
	ExecCommand = "EXEC_COMMAND",
	WaitForHttpGetEndpointAvailability = "WAIT_FOR_HTTP_GET_ENDPOINT_AVAILABILITY",
	WaitForHttpPostEndpointAvailability = "WAIT_FOR_HTTP_POST_ENDPOINT_AVAILABILITY",
	ExecuteBulkCommands = "EXECUTE_BULK_COMMANDS",
	GetServices = "GET_SERVICES",
	GetLambdas = "GET_LAMBDAS"
}
// ^^^^^^^^^^^^^^^^^^^^ Update the visitor whenever you add an enum value!!! ^^^^^^^^^^^^^^^^^^^^^^^^^^^

export namespace V0CommandType {
	export function acceptVisitor(commandType: V0CommandType, visitor: V0CommandTypeVisitor): Result<null, Error> {
		let result: Result<null, Error>;
		switch (commandType) {
			case V0CommandType.LoadLambda:
				result = visitor.visitLoadLambda();
				break;
			case V0CommandType.ExecuteLambda:
				result = visitor.visitExecuteLambda();
				break;
			case V0CommandType.RegisterService:
				result = visitor.visitRegisterService();
				break;
			case V0CommandType.GenerateFiles:
				result = visitor.visitGenerateFiles();
				break;
			case V0CommandType.LoadStaticFiles:
				result = visitor.visitLoadStaticFiles();
				break;
			case V0CommandType.StartService:
				result = visitor.visitStartService();
				break;
			case V0CommandType.RemoveService:
				result = visitor.visitRemoveService();
				break;
			case V0CommandType.Repartition:
				result = visitor.visitRepartition();
				break;
			case V0CommandType.ExecCommand:
				result = visitor.visitExecCommand();
				break;
			case V0CommandType.WaitForHttpGetEndpointAvailability:
				result = visitor.visitWaitForHttpGetEndpointAvailability();
				break;
			case V0CommandType.WaitForHttpPostEndpointAvailability:
				result = visitor.visitWaitForHttpPostEndpointAvailability();
				break;
			case V0CommandType.ExecuteBulkCommands:
				result = visitor.visitExecuteBulkCommands();
				break;
			case V0CommandType.GetServices:
				result = visitor.visitGetServices();
				break;
			case V0CommandType.GetLambdas:
				result = visitor.visitGetLambdas();
				break;
			default:
				return err(new Error("Unrecognized command type " + commandType));
		}
		if (!result.isOk()) {
			return err(result.error);
		}
		return ok(null);
	}
}

File: golang/lib/bulk_command_execution/v0_bulk_command_api/v0_command_types.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package v0_bulk_command_api

import "github.com/palantir/stacktrace"

// We provide a visitor interface here so that:
//  1) all enum cases can be exhaustively handled
//  2) any changes in the enum will result in a compile break
type V0CommandTypeVisitor interface {
	VisitLoadLambda() error
	VisitExecuteLambda() error
	VisitRegisterService() error
	VisitGenerateFiles() error
	VisitLoadStaticFiles() error
	VisitStartService() error
	VisitRemoveService() error
	VisitRepartition() error
	VisitExecCommand() error
	VisitWaitForHttpGetEndpointAvailability() error
	VisitWaitForHttpPostEndpointAvailability() error
	VisitExecuteBulkCommands() error
	VisitGetServices() error
	VisitGetLambdas() error
}

type V0CommandType string

const (
	// vvvvvvvvvvvvvvvvvvvv Update the visitor whenever you add an enum value!!! vvvvvvvvvvvvvvvvvvvvvvvvvvv
	LoadLambdaCommandType                          V0CommandType = "LOAD_LAMBDA"
	ExecuteLambdaCommandType                       V0CommandType = "EXECUTE_LAMBDA"
	RegisterServiceCommandType                     V0CommandType = "REGISTER_SERVICE"
	GenerateFilesCommandType                       V0CommandType = "GENERATE_FILES"
	LoadStaticFilesCommandType                     V0CommandType = "LOAD_STATIC_FILES"
	StartServiceCommandType                        V0CommandType = "START_SERVICE"
	RemoveServiceCommandType                       V0CommandType = "REMOVE_SERVICE"
	RepartitionCommandType                         V0CommandType = "REPARTITION"
	ExecCommandCommandType                         V0CommandType = "EXEC_COMMAND"
	WaitForHttpGetEndpointAvailabilityCommandType  V0CommandType = "WAIT_FOR_HTTP_GET_ENDPOINT_AVAILABILITY"
	WaitForHttpPostEndpointAvailabilityCommandType V0CommandType = "WAIT_FOR_HTTP_POST_ENDPOINT_AVAILABILITY"
	ExecuteBulkCommandsCommandType                 V0CommandType = "EXECUTE_BULK_COMMANDS"
	GetServicesCommandType                         V0CommandType = "GET_SERVICES"
	GetLambdasCommandType                          V0CommandType = "GET_LAMBDAS"
	// ^^^^^^^^^^^^^^^^^^^^ Update the visitor whenever you add an enum value!!! ^^^^^^^^^^^^^^^^^^^^^^^^^^^
)

func (commandType V0CommandType) AcceptVisitor(visitor V0CommandTypeVisitor) error {
	var err error
	switch commandType {
	case LoadLambdaCommandType:
		err = visitor.VisitLoadLambda()
	case ExecuteLambdaCommandType:
		err = visitor.VisitExecuteLambda()
	case RegisterServiceCommandType:
		err = visitor.VisitRegisterService()
	case GenerateFilesCommandType:
		err = visitor.VisitGenerateFiles()
	case LoadStaticFilesCommandType:
		err = visitor.VisitLoadStaticFiles()
	case StartServiceCommandType:
		err = visitor.VisitStartService()
	case RemoveServiceCommandType:
		err = visitor.VisitRemoveService()
	case RepartitionCommandType:
		err = visitor.VisitRepartition()
	case ExecCommandCommandType:
		err = visitor.VisitExecCommand()
	case WaitForHttpGetEndpointAvailabilityCommandType:
		err = visitor.VisitWaitForHttpGetEndpointAvailability()
	case WaitForHttpPostEndpointAvailabilityCommandType:
		err = visitor.VisitWaitForHttpPostEndpointAvailability()
	case ExecuteBulkCommandsCommandType:
		err = visitor.VisitExecuteBulkCommands()
	case GetServicesCommandType:
		err = visitor.VisitGetServices()
	case GetLambdasCommandType:
		err = visitor.VisitGetLambdas()
	default:
		return stacktrace.NewError("Unrecognized command type '%v'", commandType)
	}
	if err != nil {
		return stacktrace.Propagate(err, "The visitor returned an error for command type '%v'", commandType)
	}
	return nil
}
